plugins/teststeps/teststeps.go

```diff
@@ -27,53 +29,80 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	for {
-		select {
-		case target := <-ch.In:
-			if target == nil {
-				// no more targets incoming
-				return nil
+	type tgtErr struct {
+		target *target.Target
+		err    error
+	}
+	tgtErrCh := make(chan tgtErr)
+	tgtInFlight := int32(0)
+	noMoreTargetsCh := make(chan struct{})
+
+	go func(tgtInFlight *int32) {
+		defer func() {
+			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, tgtInFlight)
+		}()
+		for {
+			tgt := <-ch.In
+			log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
+
+			if tgt == nil {
+				log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
+				noMoreTargetsCh <- struct{}{}
+				return
 			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
+
 			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
+				tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
 			}()
+			atomic.AddInt32(tgtInFlight, 1)
+		}
+	}(&tgtInFlight)
 
-			select {
-			case err := <-errCh:
-				if err != nil {
-					select {
-					case ch.Err <- cerrors.TargetError{Target: target, Err: err}:
-						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, target, err)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
-				} else {
-					select {
-					case ch.Out <- target:
-						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, target)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
+	noMoreTargets := false
+	defer func() {
+		log.Debugf("%s: ForEachTarget: exiting outgoing loop, targets inflight %d, the last target received %v", pluginName, tgtInFlight, noMoreTargets)
+	}()
+	for {
+		select {
+		case te := <-tgtErrCh:
+			atomic.AddInt32(&tgtInFlight, -1)
+			if te.err != nil {
+				select {
+				case ch.Err <- cerrors.TargetError{Target: te.target, Err: te.err}:
+					log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, te.target, te.err)
+				case <-cancel:
```

Comment:
I think this will be problematic. One of the requirements from the plugins perspective is that once they receive cancel/pause, they should begin returning immediately and if we don't wait for the various `f` to actually return, we'll be hiding plugins that actually do not honor cancellation signal. The runner higher up will set a timeout for this to happen, and fail if the plugins haven't returned.

Reply:
So, I think here we should have also a wait group?

Reply:
I didn't know that we expect all the ```f``` instances should return something. Was thinking that the cancel signal should be handled in "fire and forget" fashion. Will fix.

Reply:
Do I understand right that a plugin should return errors for all "inflight" targets in case of cancelation?

Reply:
So we don't expect `f` to return anything in particular, this at the moment is not really specified, but we do expect the main `Run` function of the plugin to return, and to do so as a signal that the internal plugin logic has been interrupted.

The runner at the moment ignores any error returning from the plugin after cancellation, but it does flag a plugin as non-returning if it doesn't return within a certain timeout after the cancellation signal. 

Does it make sense?

Reply:
So we just want to detect plugins which do not respect the cancelation signal? Ok, will fix.

Reply:
Didn't manage to use a work group there, because there is no way to check remaining value in a work group without blocking. wg.Wait() would block the goroutine which consumes results from ```f``` instances and they would not be finished ever. Of course we cloud start another "outgoing" goroutine which consumes and reports results from ```f``` but I think it would make things even more complex.. Pleas let me know if you would like me to try this approach.